### ISSUE TITLE
More possibilities to filter out issues

### ISSUE THREAD

**Issue by andreasschroth** [vimeo/psalm#603]
I'd like to see more possibilities to filter out issues for a project:

- Filter by placeholders in filenames (e.g. don't show rule XYZ in files `*.Extractor.php`, regardless of the folder location).
- Filter by regex in the messages.

Because right now, this is like a blocker for us in order to be able to use it and integrate it into CI. There are just too many false-positives for us, well or let's say code we do not want to change as it is intended the way it is.

**Comment by muglug** [comment]
Regex message filtering should be very simple to implement. What’s the nature of the false-positives?

**Comment by andreasschroth** [comment]
Well, not really false-positives, as the messages are absolutely correct by themselves, but it's just intended by design the way it is. For example, let's say an `AccountExtractor` class has a method `extractCustomValueObjects()`, then it's just intended by design that the method returns different value objects and thus this method has no specific return type defined.

**Comment by andreasschroth** [comment]
@weirdan Unfortunately, that's not an option for us, as I believe it should not be necessary to add docblocks to a `normal` situation like here, if it's part of the code design that every `Extractor` class has such a method and thus no return type. I feel, this should be reserved for single occurrences, if really necessary. Anyway, it would just be another way of doing it - and I definitely prefer it over adding 100 Docblocks. `PhpStan` also offers such a way to filter out messages btw.

**Comment by andreasschroth** [comment]
Is this already implemented now and usable? Couldn't find any information about it in changelogs or documentation... Thank you. 😄

**Comment by muglug** [comment]
Ha um it’s still open, so no. But there are some more ways to filter out messages already for specific issue types e.g. `<UndefinedClass>` can be suppressed by `<referencedClass name="SomeClass" />`

**Comment by Rarst** [comment]
Seem close (tell me if I should open new issues for this), but writing extension for large upstream framework (e.g. WordPress plugin) it would be useful to filter out undefined elements by pattern.

For example:

```xml
<UndefinedClass>
    <errorLevel type="suppress">
        <referencedClass name="WP_*"/>
    </errorLevel>
</UndefinedClass>
<UndefinedFunction>
    <errorLevel type="suppress">
        <referencedFunction name="wp_*"/>
       <referencedFunction name="the_*"/>
       <referencedFunction name="get_the_*"/>
    </errorLevel>
</UndefinedFunction>
```

Advantage would be that psalm config file won't need to be constantly managed as calls are added, and issues unrelated to the framework won't be overlooked like if the issue is suppressed completely.

**Comment by weirdan** [comment]
Patterns for classes, methods and functions has since been implemented.
